bee-runtime/src/node.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{event::Bus, resource::ResHandle, worker::Worker};
+
+use bee_storage::backend::StorageBackend;
+
+use async_trait::async_trait;
+use futures::{channel::oneshot, future::Future};
+
+use std::any::Any;
+
+#[async_trait]
+pub trait Node: Send + Sized + 'static {
+    type Builder: NodeBuilder<Self>;
+    type Backend: StorageBackend;
+    type Error: std::error::Error;
+
+    async fn stop(mut self) -> Result<(), Self::Error>;
+
+    fn spawn<W, G, F>(&mut self, g: G)
+    where
+        W: Worker<Self>,
+        G: FnOnce(oneshot::Receiver<()>) -> F,
+        F: Future<Output = ()> + Send + 'static;
+
+    fn worker<W>(&self) -> Option<&W>
+    where
+        W: Worker<Self> + Send + Sync;
+
+    fn register_resource<R: Any + Send + Sync>(&mut self, res: R);
+
+    fn remove_resource<R: Any + Send + Sync>(&mut self) -> Option<R>;
+
+    #[track_caller]
+    fn resource<R: Any + Send + Sync>(&self) -> ResHandle<R>;
+
+    #[track_caller]
+    fn storage(&self) -> ResHandle<Self::Backend> {
+        self.resource()
+    }
+
+    #[track_caller]
+    fn bus(&self) -> ResHandle<Bus<'static>> {
+        self.resource()
+    }
+}
+
+#[async_trait(?Send)]
+pub trait NodeBuilder<N: Node>: Sized {
+    type Error;
```

Comment:
Should this be `std::error::Error` has the one in `Node` ?

Reply:
I'm not sure what you mean. With `std::error::Error` as a bound? Perhaps...

Reply:
Changed

Reply:
Yeah, just weird to me that NodeBuilder::Error is Error and not Node::Error

Reply:
Or the other way around